File: server/src/main/java/com/larsreimann/api_editor/transformation/BoundaryAnnotationProcessor.java
package com.larsreimann.api_editor.transformation;

import com.larsreimann.api_editor.model.*;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class BoundaryAnnotationProcessor extends AbstractPackageDataTransformer {
    @Override
    public boolean shouldVisitResultsIn(
        @NotNull AnnotatedPythonFunction pythonFunction
    ) {
        return false;
    }

    @Override
    public boolean shouldVisitEnumsIn(
        @NotNull AnnotatedPythonModule pythonModule
    ) {
        return false;
    }

    @Override
    public AnnotatedPythonParameter createNewParameter(
        @NotNull AnnotatedPythonParameter oldParameter
    ) {
        ArrayList<EditorAnnotation> annotations = new ArrayList<>();
        Boundary newBoundary = oldParameter.getBoundary();
        for (EditorAnnotation editorAnnotation : oldParameter.getAnnotations()) {
            if (editorAnnotation instanceof BoundaryAnnotation) {
                newBoundary = new Boundary(
                    ((BoundaryAnnotation) editorAnnotation).isDiscrete(),
                    ((BoundaryAnnotation) editorAnnotation).getLowerIntervalLimit(),
                    ((BoundaryAnnotation) editorAnnotation).getLowerLimitType(),
                    ((BoundaryAnnotation) editorAnnotation).getUpperIntervalLimit(),
                    ((BoundaryAnnotation) editorAnnotation).getUpperLimitType()
                );
            }
            else {
                annotations.add(editorAnnotation);
            }
        }

        return oldParameter.fullCopy(
            oldParameter.getName(),
            oldParameter.getQualifiedName(),
            oldParameter.getDefaultValue(),
            oldParameter.getAssignedBy(),
            oldParameter.isPublic(),
            oldParameter.getTypeInDocs(),
            oldParameter.getDescription(),
            annotations,
            newBoundary,
            oldParameter.getOriginalDeclaration()
        );
    }

    @Override
    public AnnotatedPythonAttribute createNewAttribute(
        @NotNull AnnotatedPythonAttribute oldAttribute
    ) {
        ArrayList<EditorAnnotation> annotations = new ArrayList<>();
        Boundary newBoundary = oldAttribute.getBoundary();
        for (EditorAnnotation editorAnnotation : oldAttribute.getAnnotations()) {
            if (editorAnnotation instanceof BoundaryAnnotation) {
                newBoundary = new Boundary(
                    ((BoundaryAnnotation) editorAnnotation).isDiscrete(),
                    ((BoundaryAnnotation) editorAnnotation).getLowerIntervalLimit(),
                    ((BoundaryAnnotation) editorAnnotation).getLowerLimitType(),
                    ((BoundaryAnnotation) editorAnnotation).getUpperIntervalLimit(),
                    ((BoundaryAnnotation) editorAnnotation).getUpperLimitType()
                );
            }
            else {
                annotations.add(editorAnnotation);
            }
        }

        return oldAttribute.fullCopy(
            oldAttribute.getName(),
            oldAttribute.getQualifiedName(),
            oldAttribute.getDefaultValue(),
            oldAttribute.isPublic(),
            oldAttribute.getTypeInDocs(),
            oldAttribute.getDescription(),
            annotations,
            newBoundary,
            oldAttribute.getOriginalDeclaration()
        );
    }
}

File: server/src/main/kotlin/com/larsreimann/api_editor/transformation/BoundaryAnnotationProcessor.kt
package com.larsreimann.api_editor.transformation

import com.larsreimann.api_editor.model.Boundary
import com.larsreimann.api_editor.model.BoundaryAnnotation
import com.larsreimann.api_editor.mutable_model.PythonPackage
import com.larsreimann.api_editor.mutable_model.PythonParameter
import com.larsreimann.modeling.descendants

/**
 * Processes and removes `@boundary` annotations.
 */
fun PythonPackage.processBoundaryAnnotations() {
    this.descendants()
        .filterIsInstance<PythonParameter>()
        .forEach { it.processBoundaryAnnotations() }
}

private fun PythonParameter.processBoundaryAnnotations() {
    this.annotations
        .filterIsInstance<BoundaryAnnotation>()
        .forEach {
            this.boundary = Boundary(
                it.isDiscrete,
                it.lowerIntervalLimit,
                it.lowerLimitType,
                it.upperIntervalLimit,
                it.upperLimitType
            )
            this.annotations.remove(it)
        }
}
